### PR TITLE
Feature/authentication

### PR DESCRIPTION
**Pull request by gerimujo496**

### CODE REVIEW

File: .env.dev [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (KostandinD) [inline .env.dev]: Please make sure to not expose sensitive data in your repo. This file should be part of your `.gitignore`.

File: src/constants/controllerPath.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export const controller_path = {
```
Comment (zeroabsolute) [inline src/constants/controllerPath.ts]: @gerimujo496 Please keep a consistent naming convention for files, use either:
- Camel-case
- Snake-case
- Underscore
Not multiple of them at the same time

File: src/constants/properties.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export enum Properties {
+  ID = 'id',
```
Comment (zeroabsolute) [inline src/constants/properties.ts:2]: @gerimujo496 What do properties represent? Are these properties related to models? If yes, I would suggest separate enums for each resource/model.

File: src/modules/email/email.service.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+export class EmailService {
+  constructor(private readonly sendGridClient: SendgridClient) {}
+
+  async sendConfirmationEmail(
+    recipient: string,
+    user: CreateUserDto,
+  ): Promise<void> {
+    const mail: MailDataRequired = {
+      to: recipient,
+
```
Comment (zeroabsolute) [inline src/modules/email/email.service.ts]: @gerimujo496 Unnecessary newline

File: src/modules/email/email.service.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      from: 'geri.mujo@softup.co',
```
Comment (zeroabsolute) [inline src/modules/email/email.service.ts]: @gerimujo496 Would be better if we defined it in an env variable (e.g. mailConfig.sender) instead of hardcoded

File: src/modules/helpers/throwError.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { HttpException, HttpStatus } from '@nestjs/common';
+
+export const throwError = (status: HttpStatus, error: string) => {
+  throw new HttpException(
```
Comment (zeroabsolute) [inline src/modules/helpers/throwError.ts]: @gerimujo496 Low priority: Looks redundant, you could simply throw the HTTPError inline

File: src/modules/user/auth.controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (zeroabsolute) [inline src/modules/user/auth.controller.spec.ts:1]: @gerimujo496 Not enough test coverage, we need several test cases to cover all authentication functionalities

File: src/modules/user/auth.controller.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  @Post(controller_path.AUTH.RESEND_EMAIL)
+  resendEmail(@Query('token') token: string) {
+    return this.authService.resendEmail(token);
+  }
+
+  @Post(controller_path.AUTH.REQUEST_RESET_PASSWORD)
+  async requestToResetPassword(@Query('email') email: string) {
+    return await this.authService.requestToResetPassword(email);
+  }
+  @Get(controller_path.AUTH.RESET_PASSWORD_FORM)
```
Comment (zeroabsolute) [inline src/modules/user/auth.controller.ts:52]: @gerimujo496 Newline required

File: src/modules/user/auth.helper.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
+import { Injectable, NotFoundException } from '@nestjs/common';
```
Comment (zeroabsolute) [inline src/modules/user/auth.helper.ts]: @gerimujo496 Usually in helper files we define simple utilities (e.g. generateToken). All other methods belong to the service layer (you have communication with the data access layer, you can do it directly in the service, no need for separate helpers

File: src/modules/user/auth.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (zeroabsolute) [inline src/modules/user/auth.service.spec.ts:1]: @gerimujo496 Unit tests for services needed

File: src/modules/user/auth.service.ts [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+  constructor(
+    private usersService: UserService,
+    private usersDal: UserDal,
+    private emailService: EmailService,
+    private authHelper: AuthHelper,
+  ) {}
+
+  async createUser(createUserDto: CreateUserDto) {
+    try {
+      console.log(process.env.TOKEN_SECRET_KEY);
```
Comment (zeroabsolute) [inline src/modules/user/auth.service.ts]: @gerimujo496 DANGEROUS: We never log secret keys, please remove this as soon as possible. Secret keys must be kept secret, hence the name

File: src/modules/user/auth.service.ts [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+      if (user) {
+        throw new UnprocessableEntityException(
+          errorMessage.EMAIL_IN_USE(email),
+        );
+      }
+      const salt = randomBytes(8).toString('hex');
+      const hash = (await scrypt(password, salt, 32)) as Buffer;
+      const result = salt + '.' + hash.toString('hex');
+
+      const registerUser: CreateUserDto = {
```
Comment (zeroabsolute) [inline src/modules/user/auth.service.ts]: @gerimujo496 Naming suggestion: userDto instead of registerUser

File: src/modules/user/auth.service.ts [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        await this.authHelper.updateConfirmationTokenAndReturnNewUser(
+          registeredUser,
+        );
+
+      const confirmationToken =
+        await this.authHelper.sendConfirmationEmail(newUser);
+
+      return confirmationToken;
+    } catch (error) {
+      throwError(error.status, error.message);
```
Comment (zeroabsolute) [inline src/modules/user/auth.service.ts]: @gerimujo496 Redundant, we just rethrow the error we don't handle it. This could be done in the controller directly

File: src/constants/properties.ts (hunk 3 above)
Comment (gerimujo496) [reply src/constants/properties.ts:2]: in seperate files or in the same file multiple enums?
